### PR TITLE
Fix three SIMD-related bugs.

### PR DESCRIPTION
**Pull request by pgavlin**
The first two bugs involve SIMD nodes that are created by the JIT
for intermediate SIMD operations. These interediate operations
are not directly represented in metadata) and may therefore use
SIMD types that are not present in metadata. For example, metadata
may only use Vector<T>, but the IR may represent some intermediate
operations using the natural vector type for the target machine.
In these cases, the JIT will not be able to derive a type handle
for the SIMD type, which was causing a couple of issues:
- `gtNewTempAssign` was generating a `GT_ASG` node instead of a
  `GT_COPYBLK` when given trees involving these operations and
  types
- `gtGetStructHandleIfPresent` was returning an invalid type handle
  when dealing with SIMD-typed `GT_IND` nodes

The fix for the former is to check for SIMD-typed trees in
`gtNewTempAssign` and `gtNewCpObjNode`; the fix for the latter is
to check for SIMD-typed trees before checking for the array index
flag on a `GT_IND` in `gtGetStructHandleIfPresent`.

The third bug is that `gtNewTempAssign` was not setting the
`lvSIMDType` flag on local vars after changing their type to a SIMD
type.

### CODE REVIEW

**Comment by pgavlin** [comment]
@CarolEidt @dotnet/jit-contrib PTAL

File: src/jit/gentree.cpp [hunk 1]
```diff
@@ -11555,6 +11591,10 @@ GenTreePtr          Compiler::gtNewTempAssign(unsigned tmp, GenTreePtr val)
         varDsc->lvType = dstTyp = genActualType(valTyp);
         if (varTypeIsGC(dstTyp))
             varDsc->lvStructGcCount = 1;
+#if FEATURE_SIMD
+        else if (varTypeIsSIMD(dstTyp))
+            varDsc->lvSIMDType = 1;
+#endif
```
Comment (CarolEidt) [inline src/jit/gentree.cpp]: CC: you should put {} around both of these clauses.

**Comment by CarolEidt** [comment]
Looks good - thanks!

File: src/jit/gentree.cpp [hunk 2]
```diff
@@ -5990,37 +5990,71 @@ GenTreeBlkOp* Compiler::gtNewCpObjNode(GenTreePtr dst,
...
+            break;
+
+        case TYP_SIMD32:
+            size = 32;
+            break;
 
-    if (size >= TARGET_POINTER_SIZE)
+        default:
+            unreached();
+        }
```
Comment (sivarv) [inline src/jit/gentree.cpp]: The above switch could be replaced with genTypeSize(type)?

**Comment by AndyAyersMS** [comment]
@pgavlin how did you run across these? Are you planning on adding some tests?

File: src/jit/gentree.cpp [hunk 3]
```diff
@@ -5990,37 +5990,71 @@ GenTreeBlkOp* Compiler::gtNewCpObjNode(GenTreePtr dst,
...
+    {
+#if FEATURE_SIMD
+        assert(src->OperGet() == GT_ADDR);
+
+        GenTree* srcValue = src->gtGetOp1();
+
+        type = srcValue->TypeGet();
+        assert(varTypeIsSIMD(type));
+
+        switch (type)
```
Comment (sivarv) [inline src/jit/gentree.cpp]: Please add a comment line describing the reason for this logic (that was mentioned in PR description)

File: src/jit/gentree.cpp [hunk 4]
```diff
@@ -11601,13 +11641,14 @@ GenTreePtr          Compiler::gtNewTempAssign(unsigned tmp, GenTreePtr val)
     // struct types. But we don't have a convenient way to do that for all SIMD temps.
 
     CORINFO_CLASS_HANDLE structHnd = gtGetStructHandleIfPresent(val);
-    if (varTypeIsStruct(valTyp) && (structHnd != NO_CLASS_HANDLE))
+    if (varTypeIsStruct(valTyp) && ((structHnd != NO_CLASS_HANDLE) || (varTypeIsSIMD(valTyp))))
```
Comment (sivarv) [inline src/jit/gentree.cpp]: Here too a comment would be useful that we permit simd type nodes with structHdn = null.

**Comment by sivarv** [comment]
Looks good - except for minor comments.

File: src/jit/gentree.cpp (hunk 2 above)
Comment (pgavlin) [inline src/jit/gentree.cpp]: Done.

File: src/jit/gentree.cpp (hunk 3 above)
Comment (pgavlin) [inline src/jit/gentree.cpp]: Done.

File: src/jit/gentree.cpp (hunk 4 above)
Comment (pgavlin) [inline src/jit/gentree.cpp]: Done.

**Comment by pgavlin** [comment]
@AndyAyersMS: I hit these when running the comma removal prototype, which hoists subtrees into assignments to temps. AFAIK existing code does not do this with SIMD trees; e.g. I was not able to trigger these issues with any of the existing tests. The best chance at reproducing this may be to try and convince CSE to replace some SIMD trees; I'll see what I can do. I've filed #6018 for follow-up.

File: src/jit/gentree.cpp (hunk 1 above)
Comment (pgavlin) [inline src/jit/gentree.cpp]: Done.

File: src/jit/gentree.cpp [hunk 5]
```diff
@@ -5970,15 +5970,15 @@ GenTreeObj* Compiler::gtNewObjNode(CORINFO_CLASS_HANDLE structHnd, GenTree* addr
...
-//  dst       - The target to copy the struct to
-//  src       - The source to copy the struct from
-//  structHnd - A class token that represents the type of object being copied.
-//  volatil   - Is this marked as volatile memory?
-
+//  dst        - The target to copy the struct to
+//  src        - The source to copy the struct from
+//  structHnd  - A class token that represents the type of object being copied. May be null
+//               if FEATURE_SIMD is enabled and the source has a SIMD type.
+//  isVolatile - Is this marked as volatile memory?
```
Comment (briansull) [inline src/jit/gentree.cpp]: I would remove the question mark from this comment.
Possibly write:
 true if this is marked as a volatile operation

**Pull request merged**